File: python/trainer/lib/test/llr.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 :
# Laurent El Shafey <laurent.el-shafey@idiap.ch>
# Sat Sep 1 9:43:00 2012 +0200
#
# Copyright (C) 2011-2012 Idiap Research Institute, Martigny, Switzerland
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""Test trainer package
"""

import os, sys
import unittest
import bob
import random
import numpy

class LLRTest(unittest.TestCase):
  """Performs various tests for Linear Logistic Regression."""
  
  def test01_llr(self):

    # Tests our LLR Trainer.
    ar1 = bob.io.Arrayset()
    ar1.append(numpy.array([1.,1.2,-1.]))
    ar1.append(numpy.array([2.,2.1,2.2]))
    ar1.append(numpy.array([3.,2.9,3.1]))
    ar1.append(numpy.array([4.,3.7,4.]))
    ar1.append(numpy.array([5.,5.2,4.9]))
    ar1.append(numpy.array([6.,6.1,5.9]))
    ar1.append(numpy.array([7.,7.,7.3]))

    ar2 = bob.io.Arrayset()
    ar2.append(numpy.array([-10.,-9.2,-1.]))
    ar2.append(numpy.array([-5.,-4.1,-0.5]))
    ar2.append(numpy.array([-10.,-9.9,-1.8]))
    ar2.append(numpy.array([-5.,-5.4,-0.3]))
    ar2.append(numpy.array([-10.,-9.3,-0.7]))
    ar2.append(numpy.array([-5.,-4.5,-0.5]))
    ar2.append(numpy.array([-10.,-9.7,-1.2]))
    ar2.append(numpy.array([-5.,-4.8,-0.2]))


    # Expected trained machine
    weights_ref= numpy.array([[13.5714], [19.3997], [-0.6432]])
    bias_ref = numpy.array([55.3255])

    # Features and expected outputs of the trained machine
    feat1 = numpy.array([1.,2.,3.])
    out1 = 105.7668
    feat2 = numpy.array([2.,3.,4.])
    out2 = 138.0947

  
    # Trains a machine (method 1)
    T = bob.trainer.LLRTrainer()
    machine1 = T.train(ar1,ar2)

    # Makes sure results are good
    self.assertTrue( (abs(machine1.weights - weights_ref) < 2e-4).all() )
    self.assertTrue( (abs(machine1.biases - bias_ref) < 2e-4).all() )
    self.assertTrue( abs(machine1(feat1) - out1) < 2e-4 )
    self.assertTrue( abs(machine1(feat2) - out2) < 2e-4 )

    # Trains a machine (method 2)
    machine2 = bob.machine.LinearMachine()
    T.train(machine2, ar1, ar2)

    # Makes sure results are good
    self.assertTrue( (abs(machine2.weights - weights_ref) < 2e-4).all() )
    self.assertTrue( (abs(machine2.biases - bias_ref) < 2e-4).all() )
    self.assertTrue( abs(machine2(feat1) - out1) < 2e-4 )
    self.assertTrue( abs(machine2(feat2) - out2) < 2e-4 )

# Instantiates our standard main module for unittests
main = bob.helper.unittest_main(LLRTest)

File: src/trainer/python/llr.cc
/**
 * @file python/trainer/src/llr.cc
 * @date Sat Sep 1 21:16:00 2012 +0200
 * @author Laurent El Shafey <laurent.el-shafey@idiap.ch>
 *
 * @brief Python bindings to Linear Logistic Regression trainer
 *
 * Copyright (C) 2011-2012 Idiap Research Institute, Martigny, Switzerland
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "bob/core/python/ndarray.h"
#include <boost/python/stl_iterator.hpp>
#include "bob/trainer/LLRTrainer.h"

using namespace boost::python;

object train1(const bob::trainer::LLRTrainer& t, 
  const bob::io::Arrayset& data1, const bob::io::Arrayset& data2) 
{
  bob::machine::LinearMachine m;
  t.train(m, data1, data2);
  return object(m);
}

void train2(const bob::trainer::LLRTrainer& t, bob::machine::LinearMachine& m, 
  const bob::io::Arrayset& data1, const bob::io::Arrayset& data2) 
{
  t.train(m, data1, data2);
}

void bind_trainer_llr() 
{
  class_<bob::trainer::LLRTrainer, boost::shared_ptr<bob::trainer::LLRTrainer> >("LLRTrainer", "Trains a linear machine to perform Linear Logistic Regression. References:\n1. A comparison of numerical optimizers for logistic regression, T. Minka, http://research.microsoft.com/en-us/um/people/minka/papers/logreg/\n2. FoCal, http://www.dsp.sun.ac.za/~nbrummer/focal/.", init<optional<const double, const double, const size_t> >((arg("prior")=0.5, arg("convergence_threshold")=1e-5, arg("max_iterations")=10000), "Initializes a new Linear Logistic Regression trainer. The training stage will place the resulting weights (and bias) in a linear machine with a single output dimension."))
    .def(init<bob::trainer::LLRTrainer&>(args("other")))
    .def(self == self)
    .def(self != self)
    .add_property("prior", &bob::trainer::LLRTrainer::getPrior, &bob::trainer::LLRTrainer::setPrior, "The synthetic prior (should be in range ]0.,1.[.")
    .add_property("convergence_threshold", &bob::trainer::LLRTrainer::getConvergenceThreshold, &bob::trainer::LLRTrainer::setConvergenceThreshold, "The convergence threshold for the conjugate gradient algorithm")
    .add_property("max_iterations", &bob::trainer::LLRTrainer::getMaxIterations, &bob::trainer::LLRTrainer::setMaxIterations, "The maximum number of iterations for the conjugate gradient algorithm")
    .def("train", &train1, (arg("self"), arg("data1"), arg("data2")), "Trains a LinearMachine to perform the Linear Logistic Regression, using two arraysets for training, one for each of the two classes (target vs. non-target). The trained LinearMachine is returned.")
    .def("train", &train2, (arg("self"), arg("machine"), arg("data1"), arg("data2")), "Trains a LinearMachine to perform the Linear Logistic Regression, using two arraysets for training, one for each of the two classes (target vs. non-target).")
    ;
}
